poclu/misc.c

```diff
@@ -113,19 +122,57 @@ poclu_get_multiple_devices (cl_platform_id *platform, cl_context *context,
   if (err != CL_SUCCESS)
     return err;
 
-  err = clGetDeviceIDs (*platform, CL_DEVICE_TYPE_ALL, 0, NULL, num_devices);
-  if (err != CL_SUCCESS)
-    return err;
+  err = clGetDeviceIDs (*platform, CL_DEVICE_TYPE_ALL, 0, NULL, &num_dev_all);
+
+  // if custom devices are excluded, continue as normal
+  if (include_custom_dev != 1)
+    {
+      if (err != CL_SUCCESS)
+        return err;
+    }
+  else
+    {
+      cl_int err_custom;
+      err_custom = clGetDeviceIDs (*platform, CL_DEVICE_TYPE_CUSTOM, 0, NULL, &num_dev_custom);
+
+      if (err != CL_SUCCESS && err_custom != CL_SUCCESS)
+        {
+          return err_custom;
+        }
+      if (err != CL_SUCCESS)
+        {
+          num_dev_all = 0;
+        }
+      if (err_custom != CL_SUCCESS)
+        {
+          num_dev_custom = 0;
+        }
+
+    }
+
+  *num_devices = num_dev_all + num_dev_custom;
 
   cl_device_id *devs
-      = (cl_device_id *)calloc (*num_devices, sizeof (cl_device_id));
+      = (cl_device_id *) calloc (*num_devices, sizeof (cl_device_id));
   cl_command_queue *ques
-      = (cl_command_queue *)calloc (*num_devices, sizeof (cl_command_queue));
+      = (cl_command_queue *) calloc (*num_devices, sizeof (cl_command_queue));
 
-  err = clGetDeviceIDs (*platform, CL_DEVICE_TYPE_ALL, *num_devices, devs,
-                        NULL);
-  if (err != CL_SUCCESS)
-    return err;
+  if (num_dev_all != 0)
+    {
+      err = clGetDeviceIDs (*platform, CL_DEVICE_TYPE_ALL, *num_devices, devs,
```

Comment:
Should it instead be the _num_dev_all_ in this query? A bit nitpicky, since using either one here might work, since the spec defines:
_The number of OpenCL devices returned is the mininum of the value specified by num_entries or the number of OpenCL devices whose type matches device_type_

Reply:
you're right, fixed